maggy/core/executors/torch_dist_executor.py

```
@@ -0,0 +1,352 @@
+    """Wraps the user supplied training function in order to be passed to the Spark Executors.
+
+    :param train_fn: Original training function.
+    :param config: Experiment config.
+    :param app_id: Maggy application ID.
+    :param run_id: Maggy run ID.
+    :param server_addr: IP of the Maggy worker registration RPC server.
+    :param hb_interval: Worker heartbeat interval.
+    :param secret: Secret string to authenticate messages.
+    :param log_dir: Location of the logger file directory on the file system.
+
+    :returns: Patched function to execute on the Spark executors.
+    """
+
+    def wrapper_function(_: Any) -> None:
+        """Patched function from dist_executor_fn factory.
+
+        :param _: Necessary catch for the iterator given by Spark to the
+        function upon foreach calls. Can safely be disregarded.
+        """
+        EnvSing.get_instance().set_ml_id(app_id, run_id)
+        partition_id, _ = util.get_partition_attempt_id()
+        client = Client(server_addr, partition_id, 0, hb_interval, secret)
+        log_file = log_dir + "/executor_" + str(partition_id) + ".log"
+
+        builtin_print = __builtin__.print
+        reporter = Reporter(log_file, partition_id, 0, builtin_print)
+
+        def maggy_print(*args, **kwargs):
+            builtin_print(*args, **kwargs)
+            reporter.log(" ".join(str(x) for x in args), True)
+
+        __builtin__.print = maggy_print
+
+        try:
+            _register_with_servers(client, reporter, partition_id)
+            tb_logdir, trial_log_file = _setup_logging(reporter, log_dir)
+            reporter.log("Awaiting worker reservations.", True)
+            client.await_reservations()
+            reporter.log("Reservations complete, configuring PyTorch.", True)
+            master_config = client.get_exec_config()[0]
+            if not master_config:
+                reporter.log(
+                    "PyTorch registration failed, exiting from all tasks.", True
```

why don't throw an exception?

---

Good catch. Should throw a runtime exception.